### ISSUE TITLE
TypeError: type 'function' is not an acceptable base type

### ISSUE THREAD

**Issue by chpeagle7** [Nuitka/Nuitka#2399]
I have written a simple test code that runs fine directly from Python without any issues. However, when it is packaged with Nuitka, an error occurs.

my code:
meta.py
```python

from typing import Tuple, Type, Callable, TypeVar, Any, Dict
import abc

_T = TypeVar("_T")
_F = TypeVar("_F", bound=Callable[..., Any])

class BaseMetaClass(abc.ABCMeta):
    def __new__(
        cls: Type[_T], clsName: str, bases: Tuple[type], attrs: Dict[str, Any]
    ) -> _T:
        result = super().__new__(cls, clsName, bases, attrs)
        return result

class sharemethod(type):

    def __new__(cls: Type[_T], func: _F) -> Type[_F]:

        clsName = func.__class__.__name__
        bases = func.__class__.__bases__
        attrs = func.__dict__
        print(super(), cls, clsName, bases, attrs)
        result = super().__new__(cls, clsName, bases, attrs)
        return result

class BaseAPIMetaClass(BaseMetaClass):

    def __new__(
        cls: Type[_T], clsName: str, bases: Tuple[type], attrs: Dict[str, Any]
    ) -> _T:
        result = super().__new__(cls, clsName, bases, attrs)
        return result

    @sharemethod
    def __str__(glob) -> str:
        return "00000"

```

main.py

```python

import sys
print('version', sys.version)
import meta

def main():
    pass
    
if __name__ == '__main__':
    main()
```

Running from python is ok：
```bash
➜  python3 main.py
version 3.11.4 (main, Jun 20 2023, 16:59:59) [Clang 14.0.3 (clang-1403.0.22.14.1)]
<super: <class 'sharemethod'>, <sharemethod object>> <class 'meta.sharemethod'> function (<class 'object'>,) {}
```

nuitka Pack command：
```bash
➜  python3 -m nuitka --onefile --standalone  main.py
Nuitka-Options:INFO: Used command line options: --onefile --standalone main.py
Nuitka:INFO: Starting Python compilation with Nuitka '1.7.10' on Python '3.11' commercial grade 'not installed'.
Nuitka:INFO: Completed Python level compilation and optimization.               
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: clang (clang).
Nuitka-Scons:INFO: Backend linking program with 7 files (no progress information available for this stage).
Nuitka-Scons:INFO: Compiled 7 C files using ccache.
Nuitka-Scons:INFO: Cached C files (using ccache) with result 'cache hit': 4
Nuitka-Scons:INFO: Cached C files (using ccache) with result 'cache miss': 3
Nuitka-Postprocessing:INFO: Creating single file from dist folder, this may take a while.
Nuitka-Onefile:INFO: Running bootstrap binary compilation via Scons.
Nuitka-Onefile:INFO: Using compression for onefile payload.
Nuitka-Onefile:INFO: Onefile payload compression ratio (25.11%) size 16259838 to 4083598.
Nuitka-Scons:INFO: Onefile C compiler: clang (clang).            
Nuitka-Scons:INFO: Onefile linking program with 1 files (no progress information available for this stage).
Nuitka-Scons:INFO: Compiled 1 C files using ccache.
Nuitka-Scons:INFO: Cached C files (using ccache) with result 'cache hit': 1
Nuitka-Onefile:INFO: Keeping onefile build directory 'main.onefile-build'.
Nuitka:INFO: Keeping dist folder 'main.dist' for inspection, no need to use it.
Nuitka:INFO: Keeping build directory 'main.build'.
Nuitka:INFO: Created binary that runs on macOS 13.0 (x86_64) or higher.
Nuitka:INFO: Successfully created 'main.bin'.
```
Run the build file main.bin(error occurred)：
```bash
➜  ./main.bin
version 3.11.4 (main, Jun 20 2023, 16:59:59) [Clang 14.0.3 (clang-1403.0.22.14.1)]
<super: <class 'sharemethod'>, <sharemethod object>> <class 'meta.sharemethod'> compiled_function (<class 'function'>,) {}
Traceback (most recent call last):
  File "/private/var/folders/9s/fj6cqr5s02l91dxls0h51lx00000gn/T/onefile_7893_1692489794_155387/main.py", line 4, in <module>
    import meta
    
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/private/var/folders/9s/fj6cqr5s02l91dxls0h51lx00000gn/T/onefile_7893_1692489794_155387/meta.py", line 26, in <module meta>
    class BaseAPIMetaClass(BaseMetaClass):
    
  File "/private/var/folders/9s/fj6cqr5s02l91dxls0h51lx00000gn/T/onefile_7893_1692489794_155387/meta.py", line 34, in BaseAPIMetaClass
    @sharemethod

  File "/private/var/folders/9s/fj6cqr5s02l91dxls0h51lx00000gn/T/onefile_7893_1692489794_155387/meta.py", line 23, in __new__
    result = super().__new__(cls, clsName, bases, attrs)

TypeError: type 'function' is not an acceptable base type
```

I have noticed a significant difference between running the statement `print(super(), cls, clsName, bases, attrs)` in the meta.py file when it is executed directly from Python versus when it is executed after being packaged.

python run output:
> <super: <class 'sharemethod'>, <sharemethod object>> <class 'meta.sharemethod'> function (<class 'object'>,) {}

The package file main.bin runs the output:
> <super: <class 'sharemethod'>, <sharemethod object>> <class 'meta.sharemethod'> compiled_function (<class 'function'>,) {}

In other words, the value of bases has changed from `(<class 'object'>,)` to `(<class 'function'>,)`.

**Comment by kayhayen** [comment]
Yes, compiled functions inherit from Python functions, to type checks stand better chances. They do not actually share much of anything. For copying there is something in the User Manual. Nuitka has a clone method.

```
   def binder(func, name):
      try:
         result = func.clone()
      except AttributeError:
         result = types.FunctionType(func.__code__, func.__globals__, name=func.__name__, argdefs=func.__defaults__, closure=func.__closure__)
         result = functools.update_wrapper(result, func)
         result.__kwdefaults__ = func.__kwdefaults__

      result.__name__ = name
      return result
```

People seem to like to copy Python functions occasionally, never sure why. However, Nuitka doesn't have that.

If this is from a PyPI package extracted, which I commend you if you did that, it would be useful to know which one, because we could use the anti-bloat engine as a workound.

As an example

```
- module-name: 'transformers.utils.doc'
  anti-bloat:
    - description: 'add support for compiled functions'
      append_plain: |
        def copy_func(f):
            try:
                g = f.clone()
            except AttributeError:
                g = types.FunctionType(f.__code__, f.__globals__, name=f.__name__, argdefs=f.__defaults__, closure=f.__closure__)
                g = functools.update_wrapper(g, f)
                g.__kwdefaults__ = f.__kwdefaults__

            return g
```

For that module, it worked to just replace append code to change the function, you might have to use replacements for that.

If you are available to do a PR, this can be fixed by adding to the yaml configuration, described here: [https://nuitka.net/doc/nuitka-package-config.html](https://nuitka.net/doc/nuitka-package-config.html)

And here is more of a walk through style post: [https://nuitka.net/posts/nuitka-package-config-kickoff.html](https://nuitka.net/posts/nuitka-package-config-kickoff.html)

Here are a example PRs for other packages:

[#2030](https://github.com/Nuitka/Nuitka/pull/2030) 
[#1993](https://github.com/Nuitka/Nuitka/pull/1993)

**Comment by kayhayen** [comment]
This doesn't seem related to a PyPI package and just bad code that I told you how to solve, no feedback, closing therefore.
